File: src/ai/scheduler/src/ActionTools.cpp
#include "scheduler/ActionTools.hpp"

ActionChoice::ActionChoice(AtomicActionPtr action /*= nullptr*/, double score /*= 0*/)
	: action(action), score(score) {}

/**
 * Compute optimal atomic action to run from now, based on time of
 * realisation and points it give.
 * 
 * This function respect the sync parameter and therefore return the
 * right action to do.
 */
ActionChoice getOptimalNextAtomic(ActionPtr action, Pose2D& robotPos) {
	// Filter actions paused or finished
	if (action->state() != ActionStatus::IDLE) {
		return ActionChoice();
	}
	
	// Try to cast as atomic action
	const auto atomic = std::dynamic_pointer_cast<AtomicAction>(action);
	if (atomic) {
		// And return it in that case
		return ActionChoice(atomic, atomic->priority(robotPos));
	}
	
	// Otherwise cast as ActionBlock
	const auto block = std::dynamic_pointer_cast<ActionBlock>(action);
	if (block) {
		ActionChoice currentChoice = ActionChoice();

		for (const auto& next : block->subactions()) {
			ActionChoice nextChoice = getOptimalNextAtomic(next, robotPos);

			// If the action has a better score
			if (nextChoice.score > currentChoice.score) {
				// We choose it !
				currentChoice = nextChoice;
			}

			// If we hit a unfinished sync action, it has to be performed
			// before anything else, so we break the loop
			if (next->isSync() && next->state() != ActionStatus::DONE) {
				break;
			}
		}

		// As this is a block, we use the block's points instead of the
		// atomic action's points
		currentChoice.score = block->priority(robotPos);
	
		return currentChoice;
	}
	
	// error: neither action block nor atomic action
	return ActionChoice();
}

File: src/ai/scheduler/include/scheduler/action_tools.py
from action_manager import Action, ActionGroup

from geometry_msgs.msg import Pose2D
from ai_msgs.msg import ActionStatus

class ActionChoice:
	def __init__(self, action=None, score = 0):
		self.action = action
		self.score = score

def get_optimal_next_atomic(action: Action, robot_pos: Pose2D):
	"""
		Compute optimal atomic action to run from now, based on time of
		realisation and points it give.

		This function respect the sync parameter and therefore return the
		right action to do.
	"""

	# Filter actions paused or finished
	if action.state != ActionStatus.IDLE:
		return ActionChoice()
	
	# Try to cast as atomic action
	if type(action) == Action:
		# And return it in that case
		return ActionChoice(action, action.priority(robot_pos))
	
	# Otherwise if group
	if isinstance(action, ActionGroup):
		# Null choice at first
		choice = ActionChoice()

		# Try all subactions
		for next in action.children:
			next_choice = get_optimal_next_atomic(next, robot_pos)

			# If the action has a better score
			if next_choice.score > choice.score:
				# We choose it !
				choice = next_choice
			

			# If we hit a unfinished action in an ordered group, it must
			# be performed in priority
			if action.type == "ordered" and next.state != ActionStatus.DONE:
				break
		

		# As this is a group, we use the group's points instead of the
		# atomic action's points
		choice.score = action.priority(robot_pos)
	
		return choice
	
	
	# error: neither action group nor atomic action
	return ActionChoice()

